File: ros2_watchdog_py/ros2_watchdog_py/node_monitor.py
from rclpy.qos import QoSProfile
from rclpy.node import Node
from rclpy.duration import Duration
from rclpy.time import Time
from rclpy.subscription import Subscription
from typing import TypeVar, List, Callable
import diagnostic_msgs
from diagnostic_msgs.msg import DiagnosticStatus, KeyValue
from tum_types_py.common import ErrorLvl
from tum_type_conversions_ros_py.common import diagnostic_level_from_type


class NodeMonitor:
    def __init__(self, node: Node) -> None:
        self.__node: Node = node
        self.__status_pub_ = self.__node.create_publisher(
            DiagnosticStatus, "/core/orchestration/" + self.__node.get_name() + "_status", 1
        )
        self.__error_items_ = {}
        self.__key_value_map_ = {}
        self.__status_msg_ = DiagnosticStatus()
        self.__status_msg_.name = self.__node.get_name()
        self.set_error_lvl("startup", ErrorLvl.ERROR)
        self.set_message("Initializing")

    def update(self) -> None:
        self.__status_msg_.level = diagnostic_level_from_type(self.get_max_error_lvl())
        self.__status_msg_.values = self.__get_key_value_msg()
        self.__status_pub_.publish(self.__status_msg_)

    def set_error_lvl(self, key: str, lvl: ErrorLvl) -> None:
        self.__error_items_[key] = lvl

    def set_message(self, message: str) -> None:
        self.__status_msg_.message = message

    def initialization_finished(self) -> None:
        self.set_error_lvl("startup", ErrorLvl.OK)
        self.set_message("Initialized")

    def report_value(self, name: str, value) -> None:
        self.__key_value_map_[name] = value

    def set_status_code(self, code: int) -> None:
        self.report_value("status_code", code)
    
    def get_status_code(self) -> None:
        return self.__key_value_map_["status_code"]

    def get_update_function(self) -> Callable:
        return self.update

    def get_max_error_lvl(self) -> ErrorLvl:
        key = max(self.__error_items_, key=self.__error_items_.get)
        self.report_value("causing_key", key)
        return self.__error_items_[key]
    
    def __get_key_value_msg(self) -> List[KeyValue]:
        key_value_list = []
        for item_key in self.__key_value_map_:
            key_value = KeyValue()
            key_value.key = item_key
            key_value.value = str(self.__key_value_map_[item_key])
            key_value_list.append(key_value)
        return key_value_list

File: ros2_watchdog_cpp/src/node_monitor.cpp
#include "ros2_watchdog_cpp/node_monitor.hpp"
namespace tam::core
{
NodeMonitor::NodeMonitor(rclcpp::Node * node, bool automatic_causing_key)
: node_(node), automatic_causing_key_(automatic_causing_key)
{
  status_pub_ = this->node_->create_publisher<diagnostic_msgs::msg::DiagnosticStatus>(
    std::string("/core/orchestration/") + std::string(this->node_->get_name()) +
      std::string("_status"),
    1);

  status_msg_.name = this->node_->get_name();
  set_error_lvl("startup", tam::types::ErrorLvl::ERROR);
  set_message("Initializing");
}
std::function<void()> NodeMonitor::get_update_function()
{
  return std::bind(&NodeMonitor::update, this);
}
int NodeMonitor::get_status_code() { return std::stoi(key_value_map_.at("status_code")); }
void NodeMonitor::update()
{
  status_msg_.level = tam::type_conversions::diagnostic_level_from_type(get_max_error_lvl());
  status_msg_.values = get_key_value_msg();
  status_pub_->publish(status_msg_);
}
void NodeMonitor::set_error_lvl(const std::string & key, const tam::types::ErrorLvl & lvl)
{
  error_items_.insert_or_assign(key, lvl);
}
void NodeMonitor::set_message(const std::string & message) { status_msg_.message = message; }
void NodeMonitor::initialization_finished()
{
  set_error_lvl("startup", tam::types::ErrorLvl::OK);
  set_message("Initialized");
}
void NodeMonitor::set_status_code(const int code) { report_value("status_code", code); }
void NodeMonitor::report_value(const std::string & name, const std::string & value)
{
  key_value_map_.insert_or_assign(name, value);
}
tam::types::ErrorLvl NodeMonitor::get_max_error_lvl()
{
  auto min = std::max_element(
    error_items_.begin(), error_items_.end(),
    [](const auto & l, const auto & r) { return l.second < r.second; });
  if (automatic_causing_key_) {
    report_value("causing_key", (*min).first);
  }
  return (*min).second;
}
std::vector<diagnostic_msgs::msg::KeyValue> NodeMonitor::get_key_value_msg()
{
  std::vector<diagnostic_msgs::msg::KeyValue> msg;
  for (const auto & item : key_value_map_) {
    diagnostic_msgs::msg::KeyValue key_value;
    key_value.key = item.first;
    key_value.value = item.second;
    msg.push_back(key_value);
  }
  return msg;
}
}  // namespace tam::core
